### DISCUSSION TITLE
noob js script order question

### DISCUSSION THREAD

**Discussion by kai-tub** [pradyunsg/furo#394] [Questions] [answered]
Hi,

I am back with a n00b question...

I am writing a sphinx-extension that depends on loading a javascript file (https://github.com/asciinema/asciinema-player).
I would love to use this theme to showcase the usage but it doesn't work as expected.

The issue seems to be the _late_ loading stage of the javascript files in the html-template.

https://github.com/pradyunsg/furo/blob/5527d4395016ddf680092d075298204f07434e82/src/furo/theme/furo/base.html#L87-L92

If I manually edit the HTML file to move the generated `<script>` tag from the body to the `<header>` section (as is the case in the `base` template) everything works as expected/tested.
As a side-node: The extension transforms the directive to an empty `<div>` tag with a unique id and adds a `<script>` tag that searches for the div with the unique id to populate it with the _video-player_. See the [asciinma-player standalone example](https://github.com/asciinema/asciinema-player#use-the-standalone-player-bundle-in-your-html-page) for more details. 

I would like to point out that I have almost zero knowledge of javascript and could be missing a very simple javascript solution.

My question boils down to: Why are the javascript files loaded so late? What was the reason to _change_ this part compared to the base layout? And how could I fix my extension to also support the furo theme?

Thanks :)

BTW: I really enjoyed the TalkPythonToMe podcast with you and the jupyter-book gang :+1:

**Comment by kai-tub** [comment]
If there is otherwise no difference, could I open a PR and move the JS part of the template to the `header` part of the main template?

**Comment by pradyunsg** [comment, marked as answer]
This was done intentionally to reduce the time to the first draw on screen, since Sphinx by default bundles over 30kB of JS on every page. Without this, Furo's pagespeed stats won't be a clean 100% (https://pagespeed.web.dev/report?url=https%3A%2F%2Fpradyunsg.me%2Ffuro%2F).

Here's StackOverflow question from 2012 about this topic: https://stackoverflow.com/questions/10994335/

What are you doing in your JS that you need it to be executed before the page DOM is loaded?

**Reply by kai-tub** [reply]
Ahh. Thank you.
Yes, that clears it up a bit.

> What are you doing in your JS that you need it to be executed before the page DOM is loaded?
Sorry, I've done a bad job of describing the problem.

From the [asciinma-player standalone example](https://github.com/asciinema/asciinema-player#use-the-standalone-player-bundle-in-your-html-page):
The extension will insert the following when the provided directive is executed in reST/md 
```html
<div id='demo'></div>
<script>
    AsciinemaPlayer.create('/demo.cast', document.getElementById('demo'));
</script>
```

What is missing is:
```html
<script src="/asciinema-player.min.js"></script>
```

My current implementation uses the default `app.add_js_file(asciinema-player.min.js)` to add the javascript file dependency to the output.
This is also covered in the [StackOverflow](https://stackoverflow.com/questions/10994335/) question you've linked.

So I guess my solution would be to _manually_ add the javascript file to the `build` output and to manually include the `<script>` tag, so that the directive will be replaced with:
```html
<div id='demo'></div>
<script src="/asciinema-player.min.js"></script>
<script>
    AsciinemaPlayer.create('/demo.cast', document.getElementById('demo'));
</script>
```

**Reply by kai-tub** [reply]
Wait, after scrolling through the Sphinx documentation, I could probably just call `app.add_js_file()` from within the directive to add the custom string with a lower priority. :thinking: 
This should give
```html
<div id='demo'></div>
```
Where the directive is and
```html
<script>
    AsciinemaPlayer.create('/demo.cast', document.getElementById('demo'));
</script>
```
right before `</body>`.

PS: Sorry for spamming

**Reply by kai-tub** [reply]
Well, ok then it would work with furo and not the default sphinx theme, because than the `<script>` tag would come before `<div id=` ...

**Reply by pradyunsg** [reply]
The solution in these situations is to have your JS not depend on the order of execution.

```js
function work() { /*...*/ }

if (document.readyState == "loading") {
  document.addEventListener("DOMContentLoaded", work);
} else {
  work();
}
```

In your case, the work function is where you'd call `AsciinemaPlayer.create`.
